api/lib/opentelemetry/context.rb

```diff
@@ -29,23 +29,44 @@ def current
         Thread.current[KEY] ||= ROOT
       end
 
-      # Sets the current context
-      #
-      # @param [Context] ctx The context to be made active
+      # @api private
       def current=(ctx)
         Thread.current[KEY] = ctx
       end
 
+      # Returns the previous context so that it can be restored
+      #
+      # @param [Context] context The new context
+      # @return [Context] prev The previous context
```

Comment:
The spec refers to this as a "Token" rather than necessarily being the previous Context. Is there value in giving ourselves that flexibility or do we not need it?
```suggestion
      # @return [Context] The previous context
```

Reply:
I'll admit I find the returning of a Token a bit unclear in the context of the detach documentation.

```
This operation is intended to help making sure the correct Context is associated with the caller's current execution unit. Users can rely on it to identify a wrong call order, i.e. trying to detach a Context that is not the current instance. In this case the operation can emit a signal to warn users of the wrong call order, such as logging an error or returning an error value.

The API MUST accept the following parameters:

A Token that was returned by a previous call to attach a Context.
```

What purpose does this token serve if we want to keep attaching and detaching balanced.  Is there a situation where you want to detach and provide a token that doesn't match the current context?

Reply:
> Is there a situation where you want to detach and provide a token that doesn't match the current context?

It helps double-check that `attach` and `detach` are balanced. These functions are meant to be low-level building blocks for higher level block-structured mechanisms. Passing an opaque Token around provides implementation flexibility while also being able to check for correct use of the functions, and report meaningful errors (oh yeah, we should do that 😄 ).

Reply:
So, why a Token instead of the Context? If we were to use an array as the Context stack instead of chaining contexts together internally, we'd likely use the array index as the Token instead of the Context itself.

Reply:
Ok so that makes it a lot clearer, I was under the impression that the token was to be used to actually retrieve a specific context instead of a means to assert that the attach/detach calls were lined up.  

I think with the fact that should allow for attaching the same context multiple times it makes sense to use an array to keep track of the context stack, and to provide the array index as token to check against.

Reply:
Changes based on discussion here 0c57bd649ff7b86319151dac48abbe332ec4a64e